### DISCUSSION TITLE
Index with Uniqueness without considering  null values

### DISCUSSION THREAD

**Discussion by sumitsharansatsangi** [isar/isar#1166] [Ideas]
Currently, When we write as follows
```dart
@Collection()
class Student {
  Id? id;
  @Index(unique: true)
  late String enrollment_number;
  @Index(unique: true)
  String? roll_number;
  late String name;
  String? father name;
}
```
The requirement is like When we are creating the record, the only required parameter is `enrollment_number` and `name`. 
`roll_number` will be assigned later to the record. So, at the time of creation of the student record it will be `null`.
We have to make enrollment_number and roll_number both unique.

When we do this, inserting second record with null value in roll_number creates `unique index violated` error.

Please add an option for uniqueness  with null allowed, and not account null values while considering as unique value.

**Comment by knpaing** [comment]
If ''null'' is allowed on multiple records of ''unique'' it wouldn't be unique anymore I believe. Also in Microsoft Sql and MySql they don't allow duplicate ''null'' in unique columns, do they?

**Reply by sumitsharansatsangi** [reply]
Mysql allows. Here, is the [reference](https://dba-presents.com/databases/mysql/274-mysql-unique-index-with-null-values).

Postgresql also allows. Upto Postgresql 14, by default multiple null values are allowed with unique indexed column. But now in Postgresql 15 , they introduce an option whether to keep multiple null values allowed  or not for unique indexed column. Here, is the [reference](https://blog.rustprooflabs.com/2022/07/postgres-15-unique-improvement-with-null)

Microsoft Sql not allows, but it also has a method to allow this. Here is the [reference](https://www.dbrnd.com/2018/03/sql-server-allow-multiple-null-values-in-unique-constraint/)

Actually, In SQL standard, it is not defined how to handle multiple null values for unique column, so each database engine has implemented in it's own way.

For Isar database, I'll suggest to go with Postgresql-15  approach, Provide an option to allow or not.
Thank you.
